src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration.Settings;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.Utilities;
+using Stratis.Bitcoin.Utilities.Extensions;
+
+namespace Stratis.Bitcoin.Consensus
+{
+    public interface IChainedHeaderValidator
+    {
+        void Validate(ChainedHeader chainedHeader);
+    }
+
+    /// <summary>
+    /// A tree structure of <see cref="ChainedHeader "/> elements.
+    /// </summary>
+    /// <remarks>
+    /// A blockchain can have multiple versions of the chain, with only one being the longest chain (or the chain with most proof).
+    /// While this is not ideal it can happen naturally or maliciously, to be able to find the best chain we have to keep track of all chains we discover.
+    /// Only one chain will represent the tip of the blockchain. 
+    /// </remarks>
+    public sealed class ChainedHeaderTree
+    {
+        private readonly Network network;
+        private readonly IChainedHeaderValidator chainedHeaderValidator;
+        private readonly ILogger logger;
+        private readonly ICheckpoints checkpoints;
+        private readonly IChainState chainState;
+        private readonly ConsensusSettings consensusSettings;
+
+        /// <summary>A special peer identifier that represents our local node.</summary>
+        internal const int LocalPeerId = -1;
+
+        /// <summary>A list of headers that represent the tips of peers.</summary>
+        private readonly Dictionary<uint256, HashSet<int>> peerIdsByTipHash;
+
+        /// <summary>A list of peer identifiers that map to the peers tip has.</summary>
+        private readonly Dictionary<int, uint256> peerTipsByPeerId;
+
+        /// <summary>An indexed collection of <see cref="ChainedHeader"/> that represents a tree of chains.</summary>
+        private readonly Dictionary<uint256, ChainedHeader> chainedHeadersByHash;
+
+        internal Dictionary<uint256, ChainedHeader> GetChainedHeadersByHash => this.chainedHeadersByHash;
```

Comment:
We should mention that these 3 guys for testing purposes only. And we should also consider replacing these 3 guys with that approach with reflection so that we don't have them at all. This was done in the block store PR for the tests to access private members.

Reply:
I agree about the reflection. There is a library for that (though we need to move it to some shared project) that allows us to exclude test-only members from the design.